Update the SDK to build against latest upstream GCC 16

(newer newlib, different toolchain name, etc.)
Index: vendor/ESP8266_RTOS_SDK/components/newlib/platform_include/sys/lock.h
===================================================================
--- /dev/null
+++ vendor/ESP8266_RTOS_SDK/components/newlib/platform_include/sys/lock.h
@@ -0,0 +1,24 @@
+#pragma once
+
+#include_next <sys/lock.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef _LOCK_T _lock_t;
+
+void _lock_init(_lock_t *plock);
+void _lock_init_recursive(_lock_t *plock);
+void _lock_close(_lock_t *plock);
+void _lock_close_recursive(_lock_t *plock);
+void _lock_acquire(_lock_t *plock);
+void _lock_acquire_recursive(_lock_t *plock);
+int _lock_try_acquire(_lock_t *plock);
+int _lock_try_acquire_recursive(_lock_t *plock);
+void _lock_release(_lock_t *plock);
+void _lock_release_recursive(_lock_t *plock);
+
+#ifdef __cplusplus
+}
+#endif
Index: vendor/ESP8266_RTOS_SDK/components/newlib/platform_include/sys/reent.h
===================================================================
--- /dev/null
+++ vendor/ESP8266_RTOS_SDK/components/newlib/platform_include/sys/reent.h
@@ -0,0 +1,18 @@
+#pragma once
+
+#define _REENT_BACKWARD_BINARY_COMPAT
+#define _REENT_SDIDINIT(_ptr)  ((_ptr)->_reserved_0)
+#define _REENT_SGLUE(_ptr)     (__sglue)
+#include_next <sys/reent.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+extern void   __sinit(struct _reent *);
+
+extern struct _glue __sglue;
+
+#ifdef __cplusplus
+}
+#endif
Index: vendor/ESP8266_RTOS_SDK/components/newlib/platform_include/time.h
===================================================================
--- vendor.orig/ESP8266_RTOS_SDK/components/newlib/platform_include/time.h
+++ vendor/ESP8266_RTOS_SDK/components/newlib/platform_include/time.h
@@ -22,8 +22,16 @@ extern "C" {
 #include_next <time.h>
 
 #define _POSIX_TIMERS 1
+#ifndef CLOCK_MONOTONIC
 #define CLOCK_MONOTONIC (clockid_t)4
+#endif
+#ifndef CLOCK_BOOTTIME
 #define CLOCK_BOOTTIME (clockid_t)4
+#endif
+
+#ifndef TIME_UTC
+#define TIME_UTC 1
+#endif
 
 int clock_settime(clockid_t clock_id, const struct timespec *tp);
 int clock_gettime(clockid_t clock_id, struct timespec *tp);
Index: vendor/ESP8266_RTOS_SDK/components/newlib/src/reent_init.c
===================================================================
--- vendor.orig/ESP8266_RTOS_SDK/components/newlib/src/reent_init.c
+++ vendor/ESP8266_RTOS_SDK/components/newlib/src/reent_init.c
@@ -23,8 +23,6 @@
 #define _STR(_s)    #_s
 #define STR(_s)     _STR(_s)
 
-static struct _reent impure_data;
-
 /* This function is not part on newlib API, it is defined in libc/stdio/local.h
  * There is no nice way to get __cleanup member populated while avoiding __sinit,
  * so extern declaration is used here.
@@ -42,14 +40,11 @@ extern void _cleanup_r(struct _reent* r)
 void esp_reent_init(struct _reent* r)
 {
     memset(r, 0, sizeof(*r));
-    r->_stdout = _GLOBAL_REENT->_stdout;
-    r->_stderr = _GLOBAL_REENT->_stderr;
-    r->_stdin  = _GLOBAL_REENT->_stdin;
-    r->__cleanup = &_cleanup_r;
-    r->__sdidinit = 1;
-    r->__sglue._next = NULL;
-    r->__sglue._niobs = 0;
-    r->__sglue._iobs = NULL;
+    _REENT_STDIN(r) = _REENT_STDIN(_GLOBAL_REENT);
+    _REENT_STDOUT(r) = _REENT_STDOUT(_GLOBAL_REENT);
+    _REENT_STDERR(r) = _REENT_STDERR(_GLOBAL_REENT);
+    _REENT_CLEANUP(r) = _REENT_CLEANUP(_GLOBAL_REENT);
+    _REENT_SDIDINIT(r) = _REENT_SDIDINIT(_GLOBAL_REENT);
 }
 
 /* only declared in private stdio header file, local.h */
@@ -59,26 +54,20 @@ extern void __sfp_lock_release(void);
 void esp_reent_cleanup(void)
 {
     struct _reent* r = __getreent();
-    /* Clean up storage used by mprec functions */
-    if (r->_mp) {
-        if (_REENT_MP_FREELIST(r)) {
-            for (int i = 0; i < _Kmax; ++i) {
-                struct _Bigint *cur, *next;
-                next = _REENT_MP_FREELIST(r)[i];
-                while (next) {
-                    cur = next;
-                    next = next->_next;
-                    free(cur);
-                }
-            }
-        }
-        free(_REENT_MP_FREELIST(r));
-        free(_REENT_MP_RESULT(r));
-    }
+	_reclaim_reent(r);
+
+    r->_emergency = NULL;
+    r->_mp = NULL;
+    r->_r48 = NULL;
+    r->_localtime_buf = NULL;
+    r->_asctime_buf = NULL;
+    r->_signal_buf = NULL;
+    r->_misc = NULL;
+    r->_cvtbuf = NULL;
 
     /* Clean up "glue" (lazily-allocated FILE objects) */
-    struct _glue* prev = &_GLOBAL_REENT->__sglue;
-    for (struct _glue* cur = _GLOBAL_REENT->__sglue._next; cur != NULL;) {
+    struct _glue* prev = &_REENT_SGLUE(_GLOBAL_REENT);
+    for (struct _glue * cur = _REENT_SGLUE(_GLOBAL_REENT)._next; cur != NULL;) {
         if (cur->_niobs == 0) {
             cur = cur->_next;
             continue;
@@ -96,21 +85,11 @@ void esp_reent_cleanup(void)
             cur = cur->_next;
             continue;
         }
-        struct _glue* next = cur->_next;
+        struct _glue * next = cur->_next;
         prev->_next = next;
         free(cur);
         cur = next;
     }
-
-    /* Clean up various other buffers */
-    free(r->_mp);
-    r->_mp = NULL;
-    free(r->_r48);
-    r->_r48 = NULL;
-    free(r->_localtime_buf);
-    r->_localtime_buf = NULL;
-    free(r->_asctime_buf);
-    r->_asctime_buf = NULL;
 }
 
 /*
@@ -120,8 +99,8 @@ int esp_newlib_init(void)
 {
     const char *default_uart_dev = "/dev/uart/" STR(CONFIG_CONSOLE_UART_NUM);
 
-    _global_impure_ptr = &impure_data;
-    esp_reent_init(_global_impure_ptr);
+    memset(&__sglue, 0, sizeof(__sglue));
+    esp_reent_init(_GLOBAL_REENT);
 
     esp_vfs_dev_uart_register();
 
@@ -156,5 +135,5 @@ err:
 
 struct _reent* __getreent()
 {
-    return _global_impure_ptr;
-}
\ No newline at end of file
+    return _impure_ptr;
+}
Index: vendor/ESP8266_RTOS_SDK/components/spi_flash/src/partition.c
===================================================================
--- vendor.orig/ESP8266_RTOS_SDK/components/spi_flash/src/partition.c
+++ vendor/ESP8266_RTOS_SDK/components/spi_flash/src/partition.c
@@ -18,6 +18,13 @@
 #include <stdio.h>
 #include <sys/lock.h>
 
+#ifdef BOOTLOADER_BUILD
+typedef _LOCK_T _lock_t;
+
+void _lock_acquire(_lock_t *plock);
+void _lock_release(_lock_t *plock);
+#endif
+
 #include "esp_attr.h"
 #include "esp_flash_data_types.h"
 #include "esp_spi_flash.h"
Index: vendor/ESP8266_RTOS_SDK/components/tcpip_adapter/tcpip_adapter_lwip.c
===================================================================
--- vendor.orig/ESP8266_RTOS_SDK/components/tcpip_adapter/tcpip_adapter_lwip.c
+++ vendor/ESP8266_RTOS_SDK/components/tcpip_adapter/tcpip_adapter_lwip.c
@@ -330,13 +330,13 @@ esp_err_t tcpip_adapter_start(tcpip_adap
         }
         memcpy(esp_netif[tcpip_if]->hwaddr, mac, NETIF_MAX_HWADDR_LEN);
         if (tcpip_if == TCPIP_ADAPTER_IF_STA) {
-            ret = esp_wifi_internal_reg_rxcb(ESP_IF_WIFI_STA, tcpip_adapter_sta_recv_cb);
+            ret = esp_wifi_internal_reg_rxcb(ESP_IF_WIFI_STA, (wifi_rxcb_t) tcpip_adapter_sta_recv_cb);
             if (ret < 0) {
                 ESP_LOGE(TAG, "TCPIP adapter bind %d error\n", ESP_IF_WIFI_STA);
                 return ESP_ERR_TCPIP_ADAPTER_INVALID_PARAMS;
             }
         } else if (tcpip_if == TCPIP_ADAPTER_IF_AP) {
-            ret = esp_wifi_internal_reg_rxcb(ESP_IF_WIFI_AP, tcpip_adapter_ap_recv_cb);
+            ret = esp_wifi_internal_reg_rxcb(ESP_IF_WIFI_AP, (wifi_rxcb_t) tcpip_adapter_ap_recv_cb);
             if (ret < 0) {
                 ESP_LOGE(TAG, "TCPIP adapter bind %d error\n", ESP_IF_WIFI_AP);
                 return ESP_ERR_TCPIP_ADAPTER_INVALID_PARAMS;
Index: vendor/ESP8266_RTOS_SDK/components/wpa_supplicant/port/include/os.h
===================================================================
--- vendor.orig/ESP8266_RTOS_SDK/components/wpa_supplicant/port/include/os.h
+++ vendor/ESP8266_RTOS_SDK/components/wpa_supplicant/port/include/os.h
@@ -21,6 +21,7 @@
 #include "esp_err.h"
 // #include "esp32/rom/ets_sys.h"
 #include "esp_libc.h"
+#include <time.h>
 
 #if CONFIG_LTM_FAST
 #define LTM_FAST
Index: vendor/ESP8266_RTOS_SDK/components/wpa_supplicant/src/rsn_supp/wpa_i.h
===================================================================
--- vendor.orig/ESP8266_RTOS_SDK/components/wpa_supplicant/src/rsn_supp/wpa_i.h
+++ vendor/ESP8266_RTOS_SDK/components/wpa_supplicant/src/rsn_supp/wpa_i.h
@@ -16,7 +16,7 @@
 #define WPA_I_H
 
 struct install_key {
-    int mic_errors_seen; /* Michael MIC errors with the current PTK */
+    int32_t mic_errors_seen; /* Michael MIC errors with the current PTK */
     int keys_cleared;
     enum wpa_alg alg;
     u8 addr[ETH_ALEN];
Index: vendor/ESP8266_RTOS_SDK/tools/cmake/toolchain-esp8266.cmake
===================================================================
--- vendor.orig/ESP8266_RTOS_SDK/tools/cmake/toolchain-esp8266.cmake
+++ vendor/ESP8266_RTOS_SDK/tools/cmake/toolchain-esp8266.cmake
@@ -1,8 +1,8 @@
 set(CMAKE_SYSTEM_NAME Generic)
 
-set(CMAKE_C_COMPILER xtensa-lx106-elf-gcc)
-set(CMAKE_CXX_COMPILER xtensa-lx106-elf-g++)
-set(CMAKE_ASM_COMPILER xtensa-lx106-elf-gcc)
+set(CMAKE_C_COMPILER xtensa-esp8266-elf-gcc)
+set(CMAKE_CXX_COMPILER xtensa-esp8266-elf-g++)
+set(CMAKE_ASM_COMPILER xtensa-esp8266-elf-gcc)
 
 set(CMAKE_C_FLAGS "-mlongcalls -Wno-frame-address -Wno-format" CACHE STRING "C Compiler Base Flags")
 set(CMAKE_CXX_FLAGS "-mlongcalls -Wno-frame-address -Wno-format" CACHE STRING "C++ Compiler Base Flags")
Index: vendor/ESP8266_RTOS_SDK/components/esp8266/CMakeLists.txt
===================================================================
--- vendor.orig/ESP8266_RTOS_SDK/components/esp8266/CMakeLists.txt
+++ vendor/ESP8266_RTOS_SDK/components/esp8266/CMakeLists.txt
@@ -15,6 +15,7 @@ if(BOOTLOADER_BUILD)
     set(ESP8266_BOOTLOADER_LIBS
         "-L ${CMAKE_CURRENT_SOURCE_DIR}/lib"
         "core"
+		"gcc"
         PARENT_SCOPE
         )
 
Index: vendor/ESP8266_RTOS_SDK/components/freertos/port/esp8266/include/freertos/portmacro.h
===================================================================
--- vendor.orig/ESP8266_RTOS_SDK/components/freertos/port/esp8266/include/freertos/portmacro.h
+++ vendor/ESP8266_RTOS_SDK/components/freertos/port/esp8266/include/freertos/portmacro.h
@@ -43,7 +43,6 @@ extern "C" {
 
 #if defined(configUSE_NEWLIB_REENTRANT) && configUSE_NEWLIB_REENTRANT == 1
 #include "esp_newlib.h"
-#define _impure_ptr _global_impure_ptr
 
 #undef _REENT_INIT_PTR
 #define _REENT_INIT_PTR(p) esp_reent_init(p)
Index: vendor/ESP8266_RTOS_SDK/components/newlib/CMakeLists.txt
===================================================================
--- vendor.orig/ESP8266_RTOS_SDK/components/newlib/CMakeLists.txt
+++ vendor/ESP8266_RTOS_SDK/components/newlib/CMakeLists.txt
@@ -13,16 +13,14 @@ idf_component_register(SRCS "${srcs}"
                     PRIV_REQUIRES esp8266 freertos)  
 
 if(CONFIG_NEWLIB_NANO_FORMAT)
-set(LIBC c_nano)
+	set(SPECS "-specs=nano.specs")
 else()
-set(LIBC c)
+	set(SPECS )
 endif()
 
-set(LIBM m)
-
 # Toolchain libraries require code defined in this component
 idf_component_get_property(newlib newlib COMPONENT_LIB)
-target_link_libraries(${COMPONENT_LIB} INTERFACE ${LIBC} ${LIBM} gcc "$<TARGET_FILE:${newlib}>")
+target_link_libraries(${COMPONENT_LIB} INTERFACE c m ${SPECS} gcc "$<TARGET_FILE:${newlib}>")
 
 # Forces the linker to include __errno from this component,
 # instead of the implementations provided by newlib.
